From 85ed40732014a76d4419fbeee73dce57c07bf751 Mon Sep 17 00:00:00 2001
From: Carmine Scarpitta <cscarpit@cisco.com>
Date: Sat, 25 Oct 2025 01:02:24 +0200
Subject: [PATCH 1/5] *: Provide interface when installing/uninstalling SRv6 uA
 SIDs

Currently, when installing SRv6 uA SIDs, FRR daemons only provide the
IPv6 nexthop address without specifying the outgoing interface. For
link-local adjacencies, this can cause the kernel to select an
incorrect interface for forwarding. To ensure packets reach the
intended destination, both the nexthop address and interface must be
provided.

This commit fixes the issue by ensuring FRR daemons provide the
outgoing interface information alongside the nexthop address when
installing SRv6 uA SIDs.

Specifically, this commit does the following:
- Extend the SRv6 seg6local context structure to include interface
  information
- Modify zebra to pass this interface data to the kernel via netlink
- Update ISIS and staticd to provide the interface when
  installing/uninstalling uA SIDs

Signed-off-by: Carmine Scarpitta <cscarpit@cisco.com>
---
 isisd/isis_zebra.c     | 2 ++
 lib/srv6.h             | 1 +
 staticd/static_zebra.c | 2 ++
 zebra/rt_netlink.c     | 5 +++++
 4 files changed, 10 insertions(+)

diff --git a/isisd/isis_zebra.c b/isisd/isis_zebra.c
index 9389d6665a74..047c5371b16a 100644
--- a/isisd/isis_zebra.c
+++ b/isisd/isis_zebra.c
@@ -1071,6 +1071,7 @@ void isis_zebra_srv6_adj_sid_install(struct srv6_adjacency *sra)
 		action = ZEBRA_SEG6_LOCAL_ACTION_END_X;
 		prefixlen = IPV6_MAX_BITLEN;
 		ctx.nh6 = sra->nexthop;
+		ctx.ifindex = sra->adj->circuit->interface->ifindex;
 		break;
 	case SRV6_ENDPOINT_BEHAVIOR_END_X_NEXT_CSID:
 		action = ZEBRA_SEG6_LOCAL_ACTION_END_X;
@@ -1078,6 +1079,7 @@ void isis_zebra_srv6_adj_sid_install(struct srv6_adjacency *sra)
 			    sra->locator->node_bits_length +
 			    sra->locator->function_bits_length;
 		ctx.nh6 = sra->nexthop;
+		ctx.ifindex = sra->adj->circuit->interface->ifindex;
 		SET_SRV6_FLV_OP(ctx.flv.flv_ops,
 				ZEBRA_SEG6_LOCAL_FLV_OP_NEXT_CSID);
 		ctx.flv.lcblock_len = sra->locator->block_bits_length;
diff --git a/lib/srv6.h b/lib/srv6.h
index 804885e5c2bf..810f867a0af1 100644
--- a/lib/srv6.h
+++ b/lib/srv6.h
@@ -121,6 +121,7 @@ struct seg6local_context {
 	struct in_addr nh4;
 	struct in6_addr nh6;
 	uint32_t table;
+	ifindex_t ifindex;
 	struct seg6local_flavors_info flv;
 	uint8_t block_len;
 	uint8_t node_len;
diff --git a/staticd/static_zebra.c b/staticd/static_zebra.c
index e210efa18421..d7cddd598cf5 100644
--- a/staticd/static_zebra.c
+++ b/staticd/static_zebra.c
@@ -706,6 +706,7 @@ void static_zebra_srv6_sid_install(struct static_srv6_sid *sid)
 				  &sid->addr, sid->attributes.ifname);
 			return;
 		}
+		ctx.ifindex = ifp->ifindex;
 		SET_SRV6_FLV_OP(ctx.flv.flv_ops, ZEBRA_SEG6_LOCAL_FLV_OP_NEXT_CSID);
 		ctx.flv.lcblock_len = sid->locator->block_bits_length;
 		ctx.flv.lcnode_func_len = sid->locator->node_bits_length;
@@ -901,6 +902,7 @@ void static_zebra_srv6_sid_uninstall(struct static_srv6_sid *sid)
 				  &sid->addr, sid->attributes.ifname);
 			return;
 		}
+		ctx.ifindex = ifp->ifindex;
 		break;
 	case SRV6_ENDPOINT_BEHAVIOR_END_PSP_USD:
 	case SRV6_ENDPOINT_BEHAVIOR_END_NEXT_CSID_PSP_USD:
diff --git a/zebra/rt_netlink.c b/zebra/rt_netlink.c
index 8508e8eca6ed..32f2d72ac669 100644
--- a/zebra/rt_netlink.c
+++ b/zebra/rt_netlink.c
@@ -479,6 +479,9 @@ parse_encap_seg6local(struct rtattr *tb,
 		ctx->nh6 = *(struct in6_addr *)RTA_DATA(
 				tb_encap[SEG6_LOCAL_NH6]);
 
+	if (tb_encap[SEG6_LOCAL_OIF])
+		ctx->ifindex = *(uint32_t *)RTA_DATA(tb_encap[SEG6_LOCAL_OIF]);
+
 	if (tb_encap[SEG6_LOCAL_TABLE])
 		ctx->table = *(uint32_t *)RTA_DATA(tb_encap[SEG6_LOCAL_TABLE]);
 
@@ -1959,6 +1962,8 @@ static bool _netlink_nexthop_encode_seg6local_info(const struct nexthop *nexthop
 			return false;
 		if (!nl_attr_put(nlmsg, buflen, SEG6_LOCAL_NH6, &ctx->nh6, sizeof(struct in6_addr)))
 			return false;
+		if (!nl_attr_put32(nlmsg, buflen, SEG6_LOCAL_OIF, ctx->ifindex))
+			return false;
 		break;
 	case ZEBRA_SEG6_LOCAL_ACTION_END_T:
 		if (!nl_attr_put32(nlmsg, buflen, SEG6_LOCAL_ACTION, SEG6_LOCAL_ACTION_END_T))

From 8d304c1b712cb7a39f1db993ec97707e19133175 Mon Sep 17 00:00:00 2001
From: Carmine Scarpitta <cscarpit@cisco.com>
Date: Sat, 25 Oct 2025 01:02:51 +0200
Subject: [PATCH 2/5] lib: Display the interface associated with uA SIDs in the
 vty output

Extend "show ipv6 route" to display the interface associated with uA SIDs.

```
router# show ipv6 route
Codes: K - kernel route, C - connected, L - local, S - static,
       O - OSPFv3, I - IS-IS, B - BGP, T - Table, v - VNC,
       V - VNC-Direct, D - SHARP, f - OpenFabric,
       t - Table-Direct,
       > - selected route, * - FIB route, q - queued, r - rejected, b - backup
       t - trapped, o - offload failure

IPv6 unicast VRF default:
...
S>* fcbb:bbbb:1:fe40::/64 [1/0] is directly connected, eth0, seg6local uA nh6 2001::2, eth0, weight 1, 00:00:35
...
````

Signed-off-by: Carmine Scarpitta <cscarpit@cisco.com>
---
 lib/srv6.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/lib/srv6.c b/lib/srv6.c
index 191652a49b91..d175794cab52 100644
--- a/lib/srv6.c
+++ b/lib/srv6.c
@@ -182,6 +182,10 @@ const char *seg6local_context2str(char *str, size_t size,
 		return str;
 
 	case ZEBRA_SEG6_LOCAL_ACTION_END_X:
+		snprintfrr(str, size, "nh6 %pI6%s, %s", &ctx->nh6, p_flavor,
+			   ifindex2ifname(ctx->ifindex, VRF_DEFAULT));
+		return str;
+
 	case ZEBRA_SEG6_LOCAL_ACTION_END_DX6:
 		snprintfrr(str, size, "nh6 %pI6%s", &ctx->nh6, p_flavor);
 		return str;

From 67665f3fd0be959dbca8c6edd0e53697b544ebfa Mon Sep 17 00:00:00 2001
From: Carmine Scarpitta <cscarpit@cisco.com>
Date: Sat, 25 Oct 2025 01:03:35 +0200
Subject: [PATCH 3/5] lib: Display the interface associated with uA SIDs in the
 json output

Extend "show ipv6 route json" to display the interface associated with uA SIDs.

```
router# show ipv6 route json
{
...
   "fcbb:bbbb:1:fe40::/64":[
      {
         "prefix":"fcbb:bbbb:1:fe40::/64",
         "prefixLen":64,
         "protocol":"static",
         "vrfId":0,
         "vrfName":"default",
         "selected":true,
         "destSelected":true,
         "distance":1,
         "metric":0,
         "installed":true,
         "table":254,
         "internalStatus":16,
         "internalFlags":9,
         "internalNextHopNum":1,
         "internalNextHopActiveNum":1,
         "internalNextHopFibInstalledNum":1,
         "nexthopGroupId":15,
         "installedNexthopGroupId":15,
         "receivedNexthopGroupId":15,
         "uptime":"00:00:40",
         "nexthops":[
            {
               "flags":3,
               "fib":true,
               "directlyConnected":true,
               "interfaceIndex":2,
               "interfaceName":"sr0",
               "active":true,
               "weight":1,
               "seg6local":{
                  "action":"uA",
                  "sidStructure":{
                     "blockLen":32,
                     "nodeLen":16,
                     "funcLen":16,
                     "argLen":0
                  }
               },
               "seg6localContext":{
                  "flavors":[

                  ],
                  "nh6":"2001::2",
                  "interfaceName":"sr0",        <==== New field
                  "interfaceIndex":2            <==== New field
               }
            }
         ]
      }
   ],
...
}
```

Signed-off-by: Carmine Scarpitta <cscarpit@cisco.com>
---
 lib/srv6.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/lib/srv6.c b/lib/srv6.c
index d175794cab52..4b33f8ff5258 100644
--- a/lib/srv6.c
+++ b/lib/srv6.c
@@ -107,6 +107,11 @@ void seg6local_context2json(const struct seg6local_context *ctx,
 	case ZEBRA_SEG6_LOCAL_ACTION_END:
 		return;
 	case ZEBRA_SEG6_LOCAL_ACTION_END_X:
+		json_object_string_addf(json, "nh6", "%pI6", &ctx->nh6);
+		json_object_string_add(json, "interfaceName",
+				       ifindex2ifname(ctx->ifindex, VRF_DEFAULT));
+		json_object_int_add(json, "interfaceIndex", ctx->ifindex);
+		return;
 	case ZEBRA_SEG6_LOCAL_ACTION_END_DX6:
 		json_object_string_addf(json, "nh6", "%pI6", &ctx->nh6);
 		return;
